sematic/graph.py

```diff
@@ -0,0 +1,463 @@
+
+        for edge in self.edges:
+            _edges_by_id[edge.id] = edge
+
+            if edge.destination_run_id is not None:
+                _edges_by_destination_id[edge.destination_run_id].append(edge)
+
+            if edge.source_run_id is not None:
+                _edges_by_source_id[edge.source_run_id].append(edge)
+
+        setattr(
+            self, make_cache_name("edges_by_destination_id"), _edges_by_destination_id
+        )
+        setattr(self, make_cache_name("edges_by_source_id"), _edges_by_source_id)
+        setattr(self, make_cache_name("edges_by_id"), _edges_by_id)
+
+    @memoized_property
+    def edges_by_destination_id(
+        self,
+    ) -> Dict[str, List[Edge]]:
+        self._populate_edge_mappings()
+        return self.edges_by_destination_id
+
+    @memoized_property
+    def edges_by_source_id(self) -> Dict[str, List[Edge]]:
+        self._populate_edge_mappings()
+        return self.edges_by_source_id
+
+    @memoized_property
+    def edges_by_id(self) -> Dict[str, Edge]:
+        self._populate_edge_mappings()
+        return self.edges_by_id
+
+    @memoized_property
+    def artifacts_by_id(self) -> Dict[str, Artifact]:
+        return {artifact.id: artifact for artifact in self.artifacts}
+
+    def input_artifacts_ready(self, run_id: str) -> bool:
+        """
+        Does run have all input artifacts ready, i.e. upstream runs have
+        resolved?
+        """
+        return all(
+            edge.artifact_id is not None
+            for edge in self.edges_by_destination_id[run_id]
+        )
+
+    def _execution_order(self, layer_run_ids: List[str]) -> List[str]:
+        """
+        For a given graph layer (all runs have the same parent_id), this will
+        return run_ids in order of execution, i.e. upstream runs first,
+        downstream runs next. This is not deterministic, as multiple runs may
+        have the same set of upstream runs, and thus can be executed in
+        parallel.
+
+        Parameters
+        ----------
+        layer_run_ids: List[str]
+            IDs of runs in the layer. All runs are epxected to have the same
+            parent_id.
+        """
+        return _sort_layer_runs(
+            layer_run_ids, _upstream_edge_filter(self.edges_by_destination_id)
+        )
+
+    def _reverse_execution_order(self, layer_run_ids: List[str]):
+        """
+        For a given graph layer (all runs have the same parent_id), this will
+        return run_ids in order of reverse execution, i.e. downstream first,
+        upstream runs next. This is not deterministic, as runs may have multiple
+        upstream runs.
+
+        Parameters
+        ----------
+        layer_run_ids: List[str]
+            IDs of runs in the layer. All runs are epxected to have the same
+            parent_id.
+        """
+        return _sort_layer_runs(
+            layer_run_ids, _downstream_edge_filter(self.edges_by_source_id)
+        )
+
+    def _run_ids_sorted_by_layer(
```

Comment:
Use classical terms?
```suggestion
    def _run_ids_breadth_first(
```

Reply:
It's more understandable as it it. I added breadth-first to the docstring.